Update deployment flow docker compose (#36545)
sourcegraph/sourcegraph: doc/admin/deploy/docker-compose/index.md

Not sure if the Docker Compose deployment type is the right for you? Learn more about the various Sourcegraph deployment types in our [Deployment overview section](../index.md).

> The rest of this page provides generic instruction for how to install Sourcegraph with Docker Compose on a server, which could be the local machine, a server on a local network or cloud-hosted server. Follow these links for cloud-specific Sourcegraph installation guides:

Reviewer: Let's use the Oxford/serial comma for lists: "the local machine, a server on a local network, or a cloud-hosted server.

- [Deploy Sourcegraph with Docker Compose on Amazon Web Services](../../deploy/docker-compose/aws.md)
- [Deploy Sourcegraph with Docker Compose on Google Cloud](../../deploy/docker-compose/google_cloud.md)
- [Deploy Sourcegraph with Docker Compose on DigitalOcean](../../deploy/docker-compose/digitalocean.md)

## Prerequisites
Complete the following tasks before deploying Sourcegraph with Docker Compose:

- Provision an appropriately configured server to host your Sourcegraph containers. Running Sourcegraph on Windows or ARM / ARM64 images is *not* supported for production deployments.
    - Determine the number and size of the repos in your environment.
    - Determine the number of users and their engagement rate with the repos. 

>     - Configure the server resources using the [resource estimator](../resource_estimator.md) to ensure it has sufficient CPUs, memory and SSD capacity.

Author: Added note.

>     - Configure the server resources using the [resource estimator](../resource_estimator.md) to ensure it has sufficient CPUs, memory and SSD capacity.

Reviewer: "sufficient CPUs, memory, and SSD capacity"

Author: I love oxford commas! Thanks for catching them.

> - Configure firewall rules to enable secure HTTPS access to the server.

Author: Changed to generic consideration of firewall ingress configuration for the server.

> - Configure access on the server to your code host, for example by using a Personal Access Token for GitHub.   

Author: Okay I changed it to refer to accessing deployment files. I left in the example for now, but will remove it when we tweak the example workflow (which currently requires a PAT to clone from the private repo)

> - Install [Docker Compose](https://docs.docker.com/compose/) on the server (minimum Docker [v20.10.0](https://docs.docker.com/engine/release-notes/#20100) and Docker Compose: [v1.29.0](https://docs.docker.com/compose/release-notes/#1290)). Docker Compose deployments should *not* be deployed with Docker Swarm

Reviewer 2: Any concerns about having the required versions as part of a sentence, instead of a bulleted list like before? Someone scanning for versions might have trouble finding this, but I don't know if that's a common use case.

- Obtain a [Sourcegraph license](https://about.sourcegraph.com/pricing/). You can run through these instructions without one, but you must obtain a license for instances of more than 10 users.

## Installation Process Overview

> The installation process is as follows:

Author: As I was editing the line numbers have changed. I'll remove the intro sentence (presume that's what you saw).

 - [Fork the Sourcegraph Reference Repository](#fork-the-sourcegraph-reference-repository). 
 - [Clone your fork of the reference repository on the server](#clone-your-fork-onto-the-server).
 - [Create a release branch on your clone](#configure-a-release-branch). 
 - [Customize the Docker-Compose yaml file to reflect your environment variables](#make-yaml-customizations). 
 - [Build and start the containers in detached mode](#run-sourcegraph). 
 - [Perform initial management operations](#management-operations)

The following sections describe each step in more detail. 

## Fork the Sourcegraph Reference Repository
The [`sourcegraph/deploy-sourcegraph-docker`](https://github.com/sourcegraph/deploy-sourcegraph-docker/) reference repository contains everything you need to install and configure a Docker Compose Sourcegraph instance, and it will make upgrades far easier. We **strongly** recommend that you create and run Sourcegraph from your own fork of the reference repository to track customizations to the [Sourcegraph Docker Compose YAML](https://github.com/sourcegraph/deploy-sourcegraph-docker/blob/master/docker-compose/docker-compose.yaml). 

> >Note: The following steps use GitHub as an example. You can fork the reference repo into your chosen code host. 

Reviewer: ```suggestion
>NOTE: The following steps use GitHub as an example. You can fork the reference repo into your chosen code host. 
```

in keeping with markdown styleguide

> > WARNING: In GitHub, forks of public repos are also public. If you plan to store secrets (SSL certificates, external Postgres credentials, etc.) within the repository you should rather create a private fork. However, a preferable approach would be to use a Secrets Management Service.  

Reviewer: ```suggestion
> WARNING: In GitHub, forks of public repos are also public. Create a private fork if you plan to store secrets (SSL certificates, external Postgres credentials, etc.) within the repository. However, a preferable approach would be to use a Secrets Management Service.  
```

1. Use the GithHub GUI to [Create a fork](https://docs.github.com/en/get-started/quickstart/fork-a-repo#forking-a-repository) of the [sourcegraph/deploy-sourcegraph-docker](https://github.com/sourcegraph/deploy-sourcegraph-docker/) reference repository.

*Alternatively*, if you are using GitHub and you want your fork to be private, create a private clone of the reference repository in your code host. This process can be performed on any machine with access to your code host:

1. Create an [empty private repository](https://docs.github.com/en/repositories/creating-and-managing-repositories/creating-a-new-repository), for example `<you/private-repository>` in GitHub.

2. Bare clone the reference repository. 

  ```bash
  git clone --bare https://github.com/sourcegraph/deploy-sourcegraph-docker/
  ```
  
3. Navigate to the bare clone and mirror push it to your private repository.

  ```bash
  cd deploy-sourcegraph-docker.git
  git push --mirror https://github.com/<you/private-repository>.git
  ```

4. Remove your local bare clone. 
  ```bash
  cd ..
  rm -rf deploy-sourcegraph-docker.git
  ```

> ## Clone your fork onto the server

Author: OK, so I've updated it to clone onto local, configure a release branch, make customisations, commit them to your release branch, then clone the release branch onto your server (Which I assume requires a PAT in our example) Please check my command for cloning the release branch.

> WARNING: The permission of your SSH / Git configuration must be set to be readable by the user in the `gitserver` container. For example, run `chmod -v -R 600 ~/path/to/.ssh` in the folder on the host machine.

> ### Git HTTP(S) authentication

Reviewer 2: I think this should stay nested under Git configuration

See ["Environment variables in Compose"](https://docs.docker.com/compose/environment-variables/) for other ways to pass these environment variables to the relevant services (including from the command line, a .env file, etc.).

## Run Sourcegraph
Now that you have configured your yaml file you can start Sourcegraph. On the production server, move to the configuration directory and run sourcegraph:

```bash

> cd deploy-sourcegraph-docker/docker-compose

Reviewer 2: `deploy-sourcegraph-docker` is the name of the repository where the deployment code has been cloned - it won't necessarily be named that on their server (especially if they've used a private repository). Is there an easy way to make that more clear?

This is the original wording so not a big deal if we don't update it now, just wanted to mention it.

Author: I presume the subdirectory will always be called docker-compose. so I changed it to that.

6. Ensure the `sourcegraph_db.out` and `codeintel_db.out` files are moved to a safe and secure location. 

#### Restore sourcegraph databases into a new environment

> The following instructions apply **only if you are restoring your databases into a new deployment** of sourcegraph ie: a new virtual machine. If you are restoring a previously running environment, see the instructions for [restoring a previously running deployment](#restoring-sourcegraph-databases-into-an-existing-environment)

Reviewer 2: ```suggestion
The following instructions apply **only if you are restoring your databases into a new deployment** of Sourcegraph (e.g.: a new virtual machine). If you are restoring a previously running environment, see the instructions for [restoring a previously running deployment](#restore-sourcegraph-databases-into-an-existing-environment)
```

7. Browse to your sourcegraph deployment, login and verify your existing configuration has been restored

> ### Restore sourcegraph databases into an existing environment

Reviewer 2: ```suggestion
### Restore Sourcegraph databases into an existing environment
```
Maybe worth doing a find-replace on all these incorrectly-lowercased sourcegraph's, there are others in the doc. You didn't put them there, I know! These docs have needed some attention for a while.
